nx/lib/nx.ex

```diff
@@ -1120,6 +1120,91 @@ defmodule Nx do
     backend.eye(%T{type: type, shape: shape, names: names}, backend_options)
   end
 
+  @doc """
+  Extract a diagonal or create a diagonal tensor.
+
+  Extracts a diagonal if given a 2d tensor, or constructs a tensor given a 1d one.
+
+  ## Examples
+
+  When given a 1d tensor:
+
+      iex> Nx.iota({4}) |> Nx.diag
+      #Nx.Tensor<
+        s64[4][4]
+        [
+          [0, 0, 0, 0],
+          [0, 1, 0, 0],
+          [0, 0, 2, 0],
+          [0, 0, 0, 3]
+        ]
+      >
+
+  Given a 2d tensor:
+
+      iex> Nx.iota({3, 3}) |> Nx.diag()
```

Comment:
@polvalente think something like this might make the docs feel very repetitive. At the same time, I also think that it will make the docs easier to reason about. How about showing the literal tensor once and the rest using iota?

Reply:
Makes sense. You could merge all tests in two, one for 2D tensors and one for 1D tensors. 

Then you could write the literal tensor and then some examples. For each case